File: app/src/main/java/com/jamieadkins/gwent/main/PresenterFactory.java
package com.jamieadkins.gwent.main;

import android.support.annotation.NonNull;

import com.jamieadkins.commonutils.mvp.BasePresenter;

/**
 * Created by jamiea on 26/12/16.
 */

public interface PresenterFactory<T extends BasePresenter> {

    /**
     * Create a new instance of a Presenter
     *
     * @return The Presenter instance
     */
    @NonNull
    T createPresenter();
}

File: app/src/main/java/com/jamieadkins/gwent/base/PresenterFactory.kt
package com.jamieadkins.gwent.base

import com.jamieadkins.commonutils.mvp2.BasePresenter

object PresenterFactory {

    private val presenters = mutableMapOf<String, BasePresenter<*>>()

    fun <V> getPresenter(tag: String, default: () -> BasePresenter<V>): BasePresenter<V> {
        return presenters.getOrPut(tag, default) as BasePresenter<V>
    }
}